File: LeetCode/15-3sum/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {

  /**
   * Finds all unique triplets in the array which gives the sum of zero.
   *
   * @param nums the input array of integers
   * @return a list of lists containing all unique triplets that sum to zero
   */
  public List<List<Integer>> threeSum(int[] nums) {
    List<List<Integer>> res = new ArrayList<>();
    
    // Sort the array
    Arrays.sort(nums);
    
    // Iterate through the array
    for (int i = 0; i < nums.length - 2; i++) {
      // If the current value is greater than zero, break from the loop
      if (nums[i] > 0) {
        break;
      }
      
      // Skip the same element to avoid duplicates
      if (i > 0 && nums[i] == nums[i - 1]) {
        continue;
      }
      
      // Call twoSumII for the current position i
      twoSumII(nums, i, res);
    }
    
    return res;
  }

  /**
   * Helper function to find pairs that sum to the target using two pointers.
   *
   * @param nums the input array of integers
   * @param i the current index in the array
   * @param res the result list to store triplets
   */
  private void twoSumII(int[] nums, int i, List<List<Integer>> res) {
    int lo = i + 1, hi = nums.length - 1;
    int target = -nums[i];
    
    while (lo < hi) {
      int sum = nums[lo] + nums[hi];
      
      if (sum < target) {
        lo++;
      } else if (sum > target) {
        hi--;
      } else {
        res.add(Arrays.asList(nums[i], nums[lo], nums[hi]));
        lo++;
        hi--;
        
        // Skip the same element to avoid duplicates in the result
        while (lo < hi && nums[lo] == nums[lo - 1]) {
          lo++;
        }
      }
    }
  }
}

File: LeetCode/15-3sum/solution.ts
/**
 * Finds all unique triplets in the array which give the sum of zero.
 *
 * @param nums - The input array of integers
 * @returns A list of lists containing all unique triplets that sum up to zero
 */
function threeSum(nums: number[]): number[][] {
  // Initialize the result array
  const result: number[][] = [];

  // Return empty result if input is null or has less than 3 elements
  if (!nums || nums.length < 3) {
    return result;
  }

  // Sort the array to facilitate the two-pointer approach and avoid duplicates
  nums.sort((a, b) => a - b);

  // Iterate through the array, treating each element as a potential pivot
  for (let i = 0; i < nums.length - 2; i++) {
    // Skip duplicate elements to avoid duplicate triplets in the result
    if (i > 0 && nums[i] === nums[i - 1]) {
      continue;
    }

    // Initialize a Set to store elements visited so far
    const visited: Set<number> = new Set();

    for (let j = i + 1; j < nums.length; j++) {
      const complement = -nums[i] - nums[j];

      // Check if the complement exists in the Set
      if (visited.has(complement)) {
        // Found a triplet that sums to zero, add it to the result
        result.push([nums[i], nums[j], complement]);

        // Skip duplicate elements to avoid duplicate triplets in the result
        while (j + 1 < nums.length && nums[j] === nums[j + 1]) {
          j++;
        }
      }

      // Add the current element to the Set
      visited.add(nums[j]);
    }
  }

  return result;
};
